app/inbound_sms/rest.py

```diff
@@ -60,3 +71,33 @@ def get_inbound_by_id(service_id, inbound_sms_id):
     message = dao_get_inbound_sms_by_id(service_id, inbound_sms_id)
 
     return jsonify(message.serialize()), 200
+
+
+@inbound_sms.route("/remove", methods=["POST"])
+def remove_inbound_sms_for_service(service_id):
+    data = request.get_json()
+    validate(data, remove_inbound_sms_for_service_schema)
+
+    archive = data["archive"]
+
+    try:
+        dao_remove_service_permission(service_id, INBOUND_SMS_TYPE)
+        dao_remove_inbound_sms_senders(service_id)
+        archive_or_release_inbound_number_for_service(service_id, archive)
```

Comment:
Here are three database operations. If the second or last transaction fails, data inconsistency may arise. For instance, if a user attempts to archive a number and dao_remove_inbound_sms_senders encounters an exception, archive_or_release_inbound_number_for_service will not be executed. Consequently, only dao_remove_service_permission will be successful, leading to a partial update and potential data corruption.

Transaction allows series of database operations are either all committed to the database or all rolled back if an error occurs. We can follow similar approach https://github.com/alphagov/notifications-api/blob/81af15c03ca0ec7557101752291d8305e658217f/app/dao/services_dao.py#L310C1-L331C1

Reply:
that's a good point actually! well spotted

Reply:
I added transaction method and refactored tests
https://github.com/alphagov/notifications-api/pull/4312/files#:~:text=service_sms_sender_dao%20import%20dao_remove_inbound_sms_senders-,with%20transaction()%3A,-dao_remove_service_permission(service_id